cranelift/codegen/meta/src/shared/legalize.rs

```diff
@@ -213,6 +216,30 @@ pub(crate) fn define(insts: &InstructionGroup, imm: &Immediates) -> TransformGro
     // embedded as part of arguments), so use a custom legalization for now.
     narrow.custom_legalize(iconst, "narrow_iconst");
 
+    narrow.legalize(
+        def!(istore32.I64(flags, a, ptr, offset)),
+        vec![
+            def!((al, ah) = isplit(a)),
+            def!(store.I32(flags, al, ptr, offset)),
+        ],
+    );
+
+    expand.legalize(
+        def!(a = sload32.I32(flags, ptr, offset)),
+        vec![
+            def!(b = load.I32(flags, ptr, offset)),
+            def!(a = sextend.I64(b)),
+        ],
+    );
```

Comment:
This seems to use an `I32` pointer (the controlling type) to load the low 32 bits of a register and then sign-extend to 64 bits (from the instruction definition). In 32-bit x86, what is this operation supposed to be doing? If there are no 64-bit registers available for this sign-extension, is `sextend.I64` going to legalize further to something else? And is that actually what you want? E.g. if bit 31 is 0, should another register be zeroed out; or if bit 31 is 1, should another register be set to all 1s?

Reply:
> is `sextend.I64` going to legalize further to something else?

Yes. In cases that I see (and I get no new legalizer failures with the `*32` instructions) `sextend.I64` is then processed by `isplit`, and the top half is usually just thrown away. I have no special insight for why Cranelift emits `sload32` instructions though.

Reply:
`sload32` instuctions are emitted for a certain WASM instruction.

Reply:
Ok, yeah I see it in `code_translator.rs`. I think a summary of our back and forth here would be helpful as a comment on these legalizations. The only other remaining issue is: why can't this be in the `widen` group? @bjorn3, is that clear to you?

Reply:
The ctrl typevar is `I32`, which matches the native pointer size on x86_32, so it requires the `expand` group. `widen` is only for ctrl typevars that are smaller than the native pointer size.

Reply:
Ah, good. Yeah, let's add that nice detail to the comment as well.